Serialize all foreign `SourceFile`s into proc-macro crate metadata

Aaron1011 opened the pull request:
Normally, we encode a `Span` that references a foreign `SourceFile` by
encoding information about the foreign crate. When we decode this
`Span`, we lookup the foreign crate in order to decode the `SourceFile`.

However, this approach does not work for proc-macro crates. When we load
a proc-macro crate, we do not deserialzie any of its dependencies (since
a proc-macro crate can only export proc-macros). This means that we
cannot serialize a reference to an upstream crate, since the associated
metadata will not be available when we try to deserialize it.

This commit modifies foreign span handling so that we treat all foreign
`SourceFile`s as local `SourceFile`s when serializing a proc-macro.
All `SourceFile`s will be stored into the metadata of a proc-macro
crate, allowing us to cotinue to deserialize a proc-macro crate without
needing to load any of its dependencies.

Since the number of foreign `SourceFile`s that we load during a
compilation session may be very large, we only serialize a `SourceFile`
if we have also serialized a `Span` which requires it.

Aaron1011 committed:
Serialize all foreign `SourceFile`s into proc-macro crate metadata

Normally, we encode a `Span` that references a foreign `SourceFile` by
encoding information about the foreign crate. When we decode this
`Span`, we lookup the foreign crate in order to decode the `SourceFile`.

However, this approach does not work for proc-macro crates. When we load
a proc-macro crate, we do not deserialzie any of its dependencies (since
a proc-macro crate can only export proc-macros). This means that we
cannot serialize a reference to an upstream crate, since the associated
metadata will not be available when we try to deserialize it.

This commit modifies foreign span handling so that we treat all foreign
`SourceFile`s as local `SourceFile`s when serializing a proc-macro.
All `SourceFile`s will be stored into the metadata of a proc-macro
crate, allowing us to cotinue to deserialize a proc-macro crate without
needing to load any of its dependencies.

Since the number of foreign `SourceFile`s that we load during a
compilation session may be very large, we only serialize a `SourceFile`
if we have also serialized a `Span` which requires it.

rust-highfive commented:
r? @matthewjasper

(rust_highfive has picked a reviewer for you, use r? to override)

Aaron1011 commented:
I think this code doesn't need to be changed, but I'm not 100% certain:

https://github.com/rust-lang/rust/blob/d8ed1b03c202d77248eb0d335062f46026fc29c4/src/librustc_middle/hir/map/collector.rs#L171-L182

When we encode a proc-macro crate, we'll end up serializing some foreign `SourceFile`s, which we won't hash. However, we will still include our dependency hashes in the crate hash, which will cause those `SourceFiles` to influence the final hash.

petrochenkov commented:
r? @petrochenkov

petrochenkov commented:
>I think this code doesn't need to be changed, but I'm not 100% certain

I'm also not sure, but the logic seems reasonable.

petrochenkov commented on src/librustc_metadata/rmeta/encoder.rs:
```diff
@@ -154,18 +168,24 @@ impl<'tcx> SpecializedEncoder<Span> for EncodeContext<'tcx> {
 
+        let source_files = self.required_source_files.as_mut().expect("Already encoded SourceMap!");
+        // Record the fact that we need to encode the data for this `SourceFile`
+        source_files.ensure(self.source_file_cache.1 + 1);
```
Nit: looks like `ensure` is called by `insert` automatically.

petrochenkov commented:
Something like this is definitely preferable to https://github.com/rust-lang/rust/pull/69976 as and end state.

r=me with the nit addressed

Aaron1011 commented:
I've removed the call to `ensure`.

@bors r=petrochenkov

bors commented:
:pushpin: Commit 3c1743cb40da105a120133fb3d041c66b8e092e7 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 3c1743cb40da105a120133fb3d041c66b8e092e7 -->
<!-- homu: {"type":"Approved","sha":"3c1743cb40da105a120133fb3d041c66b8e092e7","approver":"petrochenkov"} -->

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=33087&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33087/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33087/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux x86_64-gnu-llvm-8
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 3'
Agent machine name: 'fv-az578'
Current agent version: '2.171.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200621.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200621.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.171.1)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/292b0b6b-c065-4387-b274-4fab0b4ced19.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73706/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73706/merge:refs/remotes/pull/73706/merge
---
 ---> 31fea614d2f3
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> 4195cadf126d
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 4e90f6b48f05
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> dfa0a356d899
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling tracing v0.1.15
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-engine v0.14.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling tracing v0.1.15
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-engine v0.14.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling chalk-solve v0.14.0
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
---
...........................i........................................................................ 1900/10411
.................................................................................................... 2000/10411
......................................................i..i.......................................... 2100/10411
.................................................................................................... 2200/10411
............................................iiiii................................................... 2300/10411
.................................................................................................... 2500/10411
.................................................................................................... 2600/10411
.................................................................................................... 2700/10411
.................................................................................................... 2800/10411
---
...i................................................................................................ 5300/10411
.................................................................................................... 5400/10411
....................................i............................................................... 5500/10411
..............................i..................................................................... 5600/10411
..................................................ii.ii........i...i................................ 5700/10411
...................i................................................................................ 5900/10411
................i................................................................................... 6000/10411
..........................................................................ii........................ 6100/10411
.............i...................................................................................... 6200/10411
.............i...................................................................................... 6200/10411
.................................................................................................... 6300/10411
.................................................................................................... 6400/10411
.....................................ii...i..ii...........i......................................... 6500/10411
.................................................................................................... 6700/10411
.................................................................................................... 6800/10411
.................................................................................................... 6800/10411
........................................................................i..ii....................... 6900/10411
.................................................................................................... 7100/10411
.................................................................................................... 7200/10411
............................i....................................................................... 7300/10411
.................................................................................................... 7400/10411
---
.................................................................................................... 8300/10411
.................................................................................................... 8400/10411
..................................................................................i................. 8500/10411
.................................................................................................... 8600/10411
...................................iiiiii..iiiiii.i................................................. 8700/10411
.................................................................................................... 8900/10411
.................................................................................................... 9000/10411
.................................................................................................... 9100/10411
.................................................................................................... 9200/10411
---

10 // compile-flags: -Z span-debug -Z unpretty=expanded,hygiene
11 // check-pass
12 
- extern crate meta_macro /* 833#0 */;
+ extern crate meta_macro /* 834#0 */;
14 
15 fn main /* 406#0 */() { }


The actual stdout differed from the expected stdout.
The actual stdout differed from the expected stdout.
Actual stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro-hygiene/meta-macro-hygiene.stdout
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args proc-macro/meta-macro-hygiene.rs`
error: 1 errors occurred comparing output.
status: exit code: 0
status: exit code: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/proc-macro/meta-macro-hygiene.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro-hygiene" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-Z" "span-debug" "-Z" "unpretty=expanded,hygiene" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro-hygiene/auxiliary"
------------------------------------------
------------------------------------------
Def site: /checkout/src/test/ui/proc-macro/auxiliary/make-macro.rs:5:9: 8:10 (#3)
#![feature /* 280#0 */(prelude_import)]
#[prelude_import /* 527#1 */]
use std /* 687#1 */::prelude /* 526#1 */::v1 /* 783#1 */::*;
#[macro_use /* 404#1 */]
extern crate std /* 687#1 */;
// aux-build:make-macro.rs
// aux-build:meta-macro.rs
// edition:2018
// compile-flags: -Z span-debug -Z unpretty=expanded,hygiene


extern crate meta_macro /* 834#0 */;

fn main /* 406#0 */() { }
/*
Expansions:
Expansions:
0: parent: ExpnId(0), call_site_ctxt: #0, def_site_ctxt: #0, kind: Root
1: parent: ExpnId(0), call_site_ctxt: #0, def_site_ctxt: #0, kind: AstPass(StdImports)
2: parent: ExpnId(0), call_site_ctxt: #0, def_site_ctxt: #0, kind: Macro(Bang, "meta_macro::print_def_site")
SyntaxContexts:
SyntaxContexts:
#0: parent: #0, outer_mark: (ExpnId(0), Opaque)
#1: parent: #0, outer_mark: (ExpnId(1), Opaque)
#2: parent: #0, outer_mark: (ExpnId(1), Transparent)
#3: parent: #0, outer_mark: (ExpnId(2), Opaque)
#4: parent: #0, outer_mark: (ExpnId(2), Transparent)
#5: parent: #0, outer_mark: (ExpnId(2), SemiTransparent)

------------------------------------------
stderr:
------------------------------------------
---
thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:344:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-8/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "8.0.0" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --exclude src/tools/tidy
Build completed unsuccessfully in 1:01:31
Build completed unsuccessfully in 1:01:31
== clock drift check ==
  local time: Sat Jun 27 16:42:54 UTC 2020
  network time: Sat, 27 Jun 2020 16:42:54 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73706/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73706/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3840) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Aaron1011 commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit 27651490878c1ccd9714d7703efacf9744f5f84d has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 27651490878c1ccd9714d7703efacf9744f5f84d -->
<!-- homu: {"type":"Approved","sha":"27651490878c1ccd9714d7703efacf9744f5f84d","approver":"petrochenkov"} -->

Manishearth commented:
@bors r-

failed the rollup, likely

https://github.com/rust-lang/rust/pull/73824#issuecomment-650691210

https://dev.azure.com/rust-lang/rust/_build/results?buildId=33134&view=logs&j=1c64c3ff-4c37-56fe-a133-b407354bbfbf&t=4d85e1cd-e19c-562d-4a04-e8daca7cfc12&l=17311

<details>

```
failures:

---- [ui] ui/proc-macro/meta-macro-hygiene.rs stdout ----
diff of stdout:
#1: parent: #0, outer_mark: (ExpnId(1), Opaque)
#2: parent: #0, outer_mark: (ExpnId(1), Transparent)
#3: parent: #0, outer_mark: (ExpnId(2), Opaque)
#4: parent: #0, outer_mark: (ExpnId(2), Transparent)
#5: parent: #0, outer_mark: (ExpnId(2), SemiTransparent)
*/

------------------------------------------
stderr:
------------------------------------------

------------------------------------------



failures:
    [ui] ui/proc-macro/meta-macro-hygiene.rs

test result: FAILED. 10350 passed; 1 failed; 62 ignored; 0 measured; 0 filtered out
```

</details>

Aaron1011 commented:
@bors r+ rollup=never

It looks like one of the other PRs in the rollup affected a symbol ID in src/test/ui/proc-macro/meta-macro-hygiene.rs

bors commented:
:pushpin: Commit 27651490878c1ccd9714d7703efacf9744f5f84d has been approved by `Aaron1011`

<!-- @bors r=Aaron1011 27651490878c1ccd9714d7703efacf9744f5f84d -->
<!-- homu: {"type":"Approved","sha":"27651490878c1ccd9714d7703efacf9744f5f84d","approver":"Aaron1011"} -->

Aaron1011 commented:
@bors r- r=petrochenkov

bors commented:
:pushpin: Commit 27651490878c1ccd9714d7703efacf9744f5f84d has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 27651490878c1ccd9714d7703efacf9744f5f84d -->
<!-- homu: {"type":"Approved","sha":"27651490878c1ccd9714d7703efacf9744f5f84d","approver":"petrochenkov"} -->

Dylan-DPC commented:
@bors retry

Dylan-DPC commented:
@bors p=1

bors commented:
:hourglass: Testing commit 27651490878c1ccd9714d7703efacf9744f5f84d with merge deda619ab276e144eb7aa2e91f4e6386b1adf063...
<!-- homu: {"type":"BuildStarted","head_sha":"27651490878c1ccd9714d7703efacf9744f5f84d","merge_sha":"deda619ab276e144eb7aa2e91f4e6386b1adf063"} -->

bors commented:
:broken_heart: Test failed - [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=33248)
<!-- homu: {"type":"BuildFailed","builder_url":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=33248","builder_name":"checks-azure"} -->

Aaron1011 commented:
It looks like the `Symbol` ids are different across platforms?

Aaron1011 committed:
Normalize symbol ids to 0 in test stdout

The number of symbols we allocate (even early on) seems to be platform
dependent. We only care about hygiene for the purposes of this test,
so just set all of the symbol ids to zero

Aaron1011 commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit a7c408d6ac74f4fef732d740e95e1d2504613bb6 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov a7c408d6ac74f4fef732d740e95e1d2504613bb6 -->
<!-- homu: {"type":"Approved","sha":"a7c408d6ac74f4fef732d740e95e1d2504613bb6","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit a7c408d6ac74f4fef732d740e95e1d2504613bb6 with merge 0a838e86b9c92a9a69477434eff4a6efb87c700c...
<!-- homu: {"type":"BuildStarted","head_sha":"a7c408d6ac74f4fef732d740e95e1d2504613bb6","merge_sha":"0a838e86b9c92a9a69477434eff4a6efb87c700c"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/822192780)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/822192780","builder_name":"checks-actions"} -->

Aaron1011 committed:
Add force-host to test aux file used by proc-macro

Aaron1011 commented:
@bors r=petrochenkov

bors commented:
:pushpin: Commit 5e0396555fc0b3721c90a97e030800714c7d3d51 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 5e0396555fc0b3721c90a97e030800714c7d3d51 -->
<!-- homu: {"type":"Approved","sha":"5e0396555fc0b3721c90a97e030800714c7d3d51","approver":"petrochenkov"} -->

rust-highfive commented:
The job `x86_64-gnu-llvm-8` of your PR [failed](https://dev.azure.com/rust-lang/rust/_build/results?buildId=33331&view=logs&jobId=698b0e4a-0480-533d-81fa-f38599aa328e) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33331/logs/69), [raw log](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_apis/build/builds/33331/logs/69)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
##[section]Starting: Linux x86_64-gnu-llvm-8
##[section]Starting: Initialize job
Agent name: 'Azure Pipelines 14'
Agent machine name: 'fv-az578'
Current agent version: '2.171.1'
##[group]Operating System
16.04.6
LTS
LTS
##[endgroup]
##[group]Virtual Environment
Environment: ubuntu-16.04
Version: 20200621.1
Included Software: https://github.com/actions/virtual-environments/blob/ubuntu16/20200621.1/images/linux/Ubuntu1604-README.md
##[endgroup]
Agent running as: 'vsts'
Prepare build directory.
Set build variables.
Download all required tasks.
Download all required tasks.
Downloading task: Bash (3.171.1)
Checking job knob settings.
   Knob: AgentToolsDirectory = /opt/hostedtoolcache Source: ${AGENT_TOOLSDIRECTORY} 
   Knob: AgentPerflog = /home/vsts/perflog Source: ${VSTS_AGENT_PERFLOG} 
Start tracking orphan processes.
##[section]Finishing: Initialize job
##[section]Starting: Configure Job Name
==============================================================================
---
========================== Starting Command Output ===========================
[command]/bin/bash --noprofile --norc /home/vsts/work/_temp/56c03cb0-3770-48db-8c1d-b21148b48c91.sh

##[section]Finishing: Disable git automatic line ending conversion
##[section]Starting: Checkout rust-lang/rust@refs/pull/73706/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
---
##[command]git remote add origin https://github.com/rust-lang/rust
##[command]git config gc.auto 0
##[command]git config --get-all http.https://github.com/rust-lang/rust.extraheader
##[command]git config --get-all http.proxy
##[command]git -c http.extraheader="AUTHORIZATION: basic ***" fetch --force --tags --prune --progress --no-recurse-submodules --depth=2 origin +refs/heads/*:refs/remotes/origin/* +refs/pull/73706/merge:refs/remotes/pull/73706/merge
---
 ---> 31fea614d2f3
Step 5/8 : ENV RUST_CONFIGURE_ARGS       --build=x86_64-unknown-linux-gnu       --llvm-root=/usr/lib/llvm-8       --enable-llvm-link-shared       --set rust.thin-lto-import-instr-limit=10
 ---> Using cache
 ---> 4195cadf126d
Step 6/8 : ENV SCRIPT python2.7 ../x.py test --exclude src/tools/tidy &&            python2.7 ../x.py test src/test/mir-opt --pass=build                                   --target=armv5te-unknown-linux-gnueabi &&            python2.7 ../x.py test src/tools/tidy
 ---> 4e90f6b48f05
Step 7/8 : ENV NO_DEBUG_ASSERTIONS=1
 ---> Using cache
 ---> dfa0a356d899
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::test::CrateLibrustc" -- not in ["src/tools/tidy"]
---
Set({"src/librustc_parse_format"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_passes"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_plugin_impl"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_privacy"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_query_system"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_save_analysis"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_serialize"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_session"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
Set({"src/librustc_span"}) not skipped for "bootstrap::doc::Rustc" -- not in ["src/tools/tidy"]
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling tracing v0.1.15
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-engine v0.14.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
   Compiling rustc_parse_format v0.0.0 (/checkout/src/librustc_parse_format)
   Compiling tracing v0.1.15
   Compiling rustc_ast_pretty v0.0.0 (/checkout/src/librustc_ast_pretty)
   Compiling rustc_hir v0.0.0 (/checkout/src/librustc_hir)
   Compiling rustc_query_system v0.0.0 (/checkout/src/librustc_query_system)
   Compiling chalk-engine v0.14.0
   Compiling rustc_hir_pretty v0.0.0 (/checkout/src/librustc_hir_pretty)
   Compiling rustc_parse v0.0.0 (/checkout/src/librustc_parse)
   Compiling rustc_ast_lowering v0.0.0 (/checkout/src/librustc_ast_lowering)
---
..............................i..................................................................... 1900/10408
.................................................................................................... 2000/10408
...............................................i..i................................................. 2100/10408
.................................................................................................... 2200/10408
.....................................iiiii.......................................................... 2300/10408
.................................................................................................... 2500/10408
.................................................................................................... 2600/10408
.................................................................................................... 2700/10408
.................................................................................................... 2800/10408
---
.................................................................................................... 5300/10408
.................................................................................................... 5400/10408
.............................i...................................................................... 5500/10408
.......................i............................................................................ 5600/10408
...........................................ii.ii........i...i....................................... 5700/10408
............i....................................................................................... 5900/10408
.........i.......................................................................................... 6000/10408
...................................................................ii............................... 6100/10408
......i............................................................................................. 6200/10408
......i............................................................................................. 6200/10408
.................................................................................................... 6300/10408
.................................................................................................... 6400/10408
...............................ii...i..ii...........i............................................... 6500/10408
.................................................................................................... 6700/10408
.................................................................................................... 6800/10408
.................................................................................................... 6800/10408
...................................................................i..ii............................ 6900/10408
.................................................................................................... 7100/10408
.................................................................................................... 7200/10408
.......................i............................................................................ 7300/10408
.................................................................................................... 7400/10408
---
.................................................................................................... 8300/10408
.................................................................................................... 8400/10408
............................................................................i....................... 8500/10408
.................................................................................................... 8600/10408
..............................iiiiii..iiiiii.i...................................................... 8700/10408
.................................................................................................... 8900/10408
.................................................................................................... 9000/10408
.................................................................................................... 9100/10408
.................................................................................................... 9200/10408
---

---- [ui] ui/proc-macro/meta-macro.rs stdout ----
diff of stdout:

- Def site: $DIR/auxiliary/make-macro.rs:5:9: 8:10 (#3)
+ Def site: $DIR/auxiliary/make-macro.rs:7:9: 10:10 (#3)


The actual stdout differed from the expected stdout.
Actual stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro/meta-macro.stdout
Actual stdout saved to /checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro/meta-macro.stdout
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args proc-macro/meta-macro.rs`
error: 1 errors occurred comparing output.
status: exit code: 0
status: exit code: 0
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/src/test/ui/proc-macro/meta-macro.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--edition=2018" "-Z" "span-debug" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/proc-macro/meta-macro/auxiliary"
------------------------------------------
------------------------------------------
Def site: /checkout/src/test/ui/proc-macro/auxiliary/make-macro.rs:7:9: 10:10 (#3)
------------------------------------------
stderr:
------------------------------------------

---
thread 'main' panicked at 'Some tests failed', src/tools/compiletest/src/main.rs:344:22
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/ui" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--mode" "ui" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-8/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0 -Zunstable-options  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python2.7" "--lldb-python" "/usr/bin/python2.7" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "8.0.0" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap test --exclude src/tools/tidy
Build completed unsuccessfully in 1:11:48
Build completed unsuccessfully in 1:11:48
== clock drift check ==
  local time: Tue Jun 30 19:59:06 UTC 2020
  network time: Tue, 30 Jun 2020 19:59:06 GMT
== end clock drift check ==

##[error]Bash exited with code '1'.
##[section]Finishing: Run build
##[section]Starting: Checkout rust-lang/rust@refs/pull/73706/merge to s
Task         : Get sources
Description  : Get sources from a repository. Supports Git, TfsVC, and SVN repositories.
Version      : 1.0.0
Author       : Microsoft
Author       : Microsoft
Help         : [More Information](https://go.microsoft.com/fwlink/?LinkId=798199)
==============================================================================
Cleaning any cached credential from repository: rust-lang/rust (GitHub)
##[section]Finishing: Checkout rust-lang/rust@refs/pull/73706/merge to s
Cleaning up task key
Start cleaning up orphan processes.
Terminate orphan process: pid (3793) (python)
##[section]Finishing: Finalize Job
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

Aaron1011 commented:
Forgot to bless a stderr file

@bors r=petrochenkov

bors commented:
:pushpin: Commit 37a48fa838e73f2c008bbbbd0db555f4852be354 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 37a48fa838e73f2c008bbbbd0db555f4852be354 -->
<!-- homu: {"type":"Approved","sha":"37a48fa838e73f2c008bbbbd0db555f4852be354","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 37a48fa838e73f2c008bbbbd0db555f4852be354 with merge 83c62c0537052bce30041002fc57bc73c98b96cf...
<!-- homu: {"type":"BuildStarted","head_sha":"37a48fa838e73f2c008bbbbd0db555f4852be354","merge_sha":"83c62c0537052bce30041002fc57bc73c98b96cf"} -->

Manishearth commented:
@bors r- yield retry

sorry, _just_ made a rollup

Manishearth commented:
@bors r=petrochenkov retry

bors commented:
:pushpin: Commit 37a48fa838e73f2c008bbbbd0db555f4852be354 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 37a48fa838e73f2c008bbbbd0db555f4852be354 -->
<!-- homu: {"type":"Approved","sha":"37a48fa838e73f2c008bbbbd0db555f4852be354","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 37a48fa838e73f2c008bbbbd0db555f4852be354 with merge d462551a8600e57d8b6f87e71ea56868bc5da6cf...
<!-- homu: {"type":"BuildStarted","head_sha":"37a48fa838e73f2c008bbbbd0db555f4852be354","merge_sha":"d462551a8600e57d8b6f87e71ea56868bc5da6cf"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/824932007), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=33364)
Approved by: petrochenkov
Pushing d462551a8600e57d8b6f87e71ea56868bc5da6cf to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=33364","checks-actions":"https://github.com/rust-lang-ci/rust/runs/824932007"},"merge_sha":"d462551a8600e57d8b6f87e71ea56868bc5da6cf"} -->

mati865 commented:
Nice performance improvement :tada:: https://perf.rust-lang.org/compare.html?start=16957bd4d3a5377263f76ed74c572aad8e4b7e59&end=d462551a8600e57d8b6f87e71ea56868bc5da6cf&stat=instructions:u
